# Assignment 5: Association mining

## Objective of this assignment
The overall objective is to understand how frequent itemsets can be extracted by
the Apriori algorithm and be able to calculate and interpret association rules in terms of support and confidence.

## ** Important: ** When handing in your homework:
+ Hand in the notebook (and nothing else) named as follows: StudentName1_snumber_StudentName2_snumber.ipynb
+ Provide clear and complete answers to the questions below under a separate header (not hidden somewhere in your source code), and make sure to explain your answers / motivate your choices. Add Markdown cells where necessary.
+ Source code, output graphs, derivations, etc., should be included in the notebook.
+ Hand-in: upload to Blackboard.
+ Include name, student number, assignment (especially in filenames)!
+ When working in pairs only one of you should upload the assignment, and report the name of your partner in your filename.
+ For problems or questions: use the BB discussion board or email the student assistants.


## Advised Reading and Exercise Material
**The following reading material is recommended:**

- Pang-Ning Tan, Michael Steinbach, and Vipin Kumar, *Introduction to Data Mining*, section 6.


## Additional Tools
For this exercise you will need to load the provided *apriorimining.py* script. 


##  5.1 Association mining for course data 
We will use the Apriori algorithm to automatically mine for associations. The Apriori algorithm is an adapted version of the script found here: https://github.com/nalinaksh/Association-Rule-Mining-Python

Check out the script and doc and check if you understand how the association rules are computed. 



In [2]:
import Toolbox.apriorimining



#### 5.1.1

(0 points) Look at the data file `Data/courses.txt` into Python. The data is represented in Table 1. Inspect the file Data/courses.txt and make sure you understand how the data in Table 1 is stored in the text file.

##### Table 1
|#  |   History |Math| Biology| Spanish | Economics| Physics | Chemistry | English  |  
| :-------------: |:-------------:| :-----------:| :----------:| :------------:|:-------------:| :------------:|  :-------------: | :-------------: |
|student 1 | 0| 1 | 0 | 0 | 1| 1 |1 |1   
|student 2 | 1| 1 | 1 | 0 | 0| 1 |1 |1   
|student 3 | 0| 1 | 0 | 1 | 0| 1 |0 |1   
|student 4 | 0| 0 | 1 | 0 | 0| 1 |1 |0   
|student 5 | 0| 1 | 0 | 0 | 0| 1 |1 |0        
|student 6 | 0| 1 | 1 | 0 | 0| 1 |1 |1   


#### 5.1.2
(1 point) We will analyze the data in Table 1 automatically using the function `associationmining.generate_association_rules()` from the script. Analyze the data with $ minsupport  \geq 80 \% $ and $ minconfidence \geq 100 \%$.What
are the generated association rules? What kind of conclusions can you make based on these association rules about the subjects that students took?  
  


In [ ]:
import Toolbox.apriorimining

Toolbox.apriorimining.generate_association_rules()

#generated association rules: 
#Rule #1: {  } ==> { 6 }
#Rule #2: { 2 } ==> { 6 }
#Rule #3: { 7 } ==> { 6 }



  ##  5.2 Association mining for MovieLens data 
  
  
  In this part of the exercise we consider a Market Basket data set containing 943 users purchases of 1682 movies. A total of 100,000 movies
have been purchased.The data set is called MovieLens100K and is provided by http://www.grouplens.org/node/73, see also the readme `MovieLensData.txt` in the data folder. The data currently considered is not the original data but modified for the apriori algorithm.

#### 5.2.1
  (0 points) The MovieLens data is stored in the file MovieLensData.txt. Inspect the file to see how the data is stored.


#### 5.2.2 
  (1 point) Find association rules using the function below with $ minsupport  \geq 30 \% $ and $ minconfidence \geq 80 \%$. What are the associations with strongest confidence? Do these associations make sense? You can use file Data/u.item to print the movie titles in stead of numbers. If you enter filename `MovieLensData.txt`, the script will provide an additional option for this. 
  

In [3]:
Toolbox.apriorimining.generate_association_rules()

Please enter support value in %: 30
Please enter confidence value in %: 80
Enter the max number of rules you want to see (enter 0 to see all rules): 0
Please enter filepath\filename and extension: Data\MovieLensData.txt
Do you want to print sets and rules with Movie names in stead of numbers? [y/n]: y
---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { Star Wars (1977) },  sup= 61.82
set= { Contact (1997) },  sup= 53.98
set= { Fargo (1996) },  sup= 53.87
set= { Return of the Jedi (1983) },  sup= 53.76
set= { Liar Liar (1997) },  sup= 51.43
set= { English Patient, The (1996) },  sup= 51.01
set= { Scream (1996) },  sup= 50.69
set= { Toy Story (1995) },  sup= 47.93
set= { Air Force One (1997) },  sup= 45.71
set= { Independence Day (ID4) (1996) },  sup= 45.49
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { Return of the Jedi (1983), Star Wars (1977) },  sup= 50.0
set= { Fargo (1

Assosiations with strongest confidence:
Rule #73: { Empire Strikes Back, The (1980), Raiders of the Lost Ark (1981), Return of the Jedi (1983) } ==> { Star Wars (1977) }, sup= 31.18, conf= 99.66

Rule #34: { Empire Strikes Back, The (1980), Return of the Jedi (1983) } ==> { Star Wars (1977) }, sup= 33.40, conf= 99.37

Rule #54: { Pulp Fiction (1994), Return of the Jedi (1983) } ==> { Star Wars (1977) }, sup= 30.22, conf= 98.62

Yes they make sense, seeing as their genres are related and "the empire strikes back", "Return of the Jedi" and "Star Wars" are all part of Star Wars. 

#### 5.2.3 
(1 point) Which movie has been bought by the most users? There are only few rules with more than three items. Why?

Star Wars has been bought by the most users. As the number of items increases, the support can only stay the same or decrease, where decrease is more likely. This is because all the three item groups automatically have the two-item groups in them since every occurence or the three-item group is also an occurence of the two-item group and there can be more two-item group occurrences where there are no three-group occurrences. The threshold for support for generating rules for groups is set at 30 and apperently there are only a few three-item groups that occur together often enough to have a support greater than 30.

#### 5.2.4
(0.5 points) Often we are interested in rules with high confidence. Is it possible for
itemsets to have very low support but still have a very high confidence?

Support stand for the fraction of transactions that containboth X and Y.
Confidence measures how often items in Y appear in transactions that contain X.
So yes it's possible for itemsets to have very low support, but still high confidence. In this case there should be few transactions that contain both X and Y, compared to the total number of transactions (low support), but X should not occur often without Y occurring -> then Y appears relatively often in transactions that contain X (high confidence).

## 5.3 Calculating support, confidence and interest

Calculate these measures and write down how you computed things, not just the answers. 


#### 5.3.1
 Suppose we have market basket data consisting of 100 transactions and 20 items. The support for item $ \text{a} = 45 \%$, the support for item $ \text{b} = 80 \%$ and the support for itemset $ \text{ {a,b }} = 30 \%$. Let the support and confidence thresholds be 20$ \%$ and 60$ \%$, respectively.
  
1. (0.5 points) Compute the confidence of the association rule $ \text{ {a } } \rightarrow   \text{{b }} $. Is the rule interesting according to the confidence measure?

2. (0.5 points) Compute the interest measure (or lift, see slide 44 of chapter 6) for the association pattern $ \text{ {a,b}}$. Describe the nature of the relationship between item $ \text{a}$ and item $ \text{b}$  in terms of the itemset measure.
3. (1 points) What conclusion can you draw from the results of parts (1) and (2)?

4. (1.5 points) Prove that if the confidence of the rule $ \text{ {a } } \rightarrow   \text{{b }} $ is less than the support of $ \text{ {b }}$  then
$$
c(  \text{ {~a } } \rightarrow   \text{{b }})> c(  \text{ {a } } \rightarrow   \text{{b }})
$$
and
$$
c(  \text{ {~a } } \rightarrow   \text{{b }})> s( {b })
$$

where $ \text{c(.)}$  denotes the rule confidence and  $ \text{s(.)}$ denote the support of an itemset. 

Hint: To prove the statement rewrite the confidence and the support of the rule in terms of probabilities: 
$$
c(  \text{ {a } } \rightarrow   \text{{b }})=  \frac{P( {a,b })}{P( {a })}
$$

and

$$
s( {b })=P( {b })
$$

Further more, make use of probability rules such as $p(b) = p(\text{~}a,b) + p(a,b)$ and $p(\text{~}a) = 1 - p(a)$ and algebraic rules such as $$ \frac{p(a,b)}{p(b)} < p(a) \rightarrow 1 - \frac{p(a,b)}{p(b)} > 1 - p(a) $$


support = x/t=support
where x is the transactions that contain x and t is the total number of transactions

1. 
Transactions that contain a: 0.45*100= 45
Transactions that contain b: 0.8*100= 80
Transactions that contain {a,b}: 0.3*100= 30
Confidence = {a,b}/a*100 = 30/45*100 = 67%
This is more than 60 so it's an interesting rule.

2.
lift = P(Y|X)/P(Y) = 0.67/0.8 = 0.83.
Probability is equal to support as in P(a)=support of a and P(b)= support of b
P(b) = 0.8
P(b|a) = P(a,b)/P(a) = 30/45 = 0.67
Lift is 83%
The probability of the rule a->b being valid measured as a ratio of strength of said rule devided by the strength of a->b if it was not a rule and transactions were distributed randomly is 83%.

3.
The rule is interesting according to the confidence measure and it has a pretty high lift. We can conclude that this is a solid rule. 
4.

Confidence of a->b = P(a,b)/P(a) 
Support of b = P(b)

If c(a->b) is smaller than s(b) then theres relatively more occurrences of b in total than b in the transactions that contain a.
Since we assume p(a,b)/p(a) < p(b) we know that 1-P(a,b)/P(a) > 1 - p(b) (using the rule p(a,b)p(b)<p(a)→1−p(a,b)p(b)>1−p(a)) then we know 1-P(a,b)/P(b) > P(¬b) (using the rule   p(~a)=1−p(a)), we then know p(¬a,b)/P(b) <  P(¬b) (since: p(b)=p(¬a,b)+p(a,b) so p(¬a,b)/p(b)+p(a,b)/p(b) = P(b)/P(b) = 1) and thus P(¬a,b)/P(¬a) > P(b) (using the rule p(a,b)p(b)<p(a)→1−p(a,b)p(b)>1−p(a)) which is equal to c( {~a } →{b })>s(b) from this we can conclude P(¬a,b)/P(¬a) > P(a,b)/P(a) which is equal to c( {¬a } →{b })>c( {a } →{b }) (since p(a,b)/p(a) < p(b) and P(¬a,b)/P(¬a) > P(b))







#### 5.3.2

(3 points) Consider the relationships between customers who buy high-definition televisions and exercise machines as shown in Table 2 and 3.

1. Compute the odd ratios for both tables.
2. Compute the $\phi$-coefficient for both tables.
3. Compute the interest factor for both tables.

For Table 3 you should compute measures given above separately for College
Students and for Adults. For each of the measures, describe how the direction
of association changes when data is pooled together (Table 2) instead of being
separated into two groups (Table 3)

##### Table 2: Two way contingency table between the sale of high-definition television and exercise machine
| |   Buy Exercise machine |     |     |
| :------------- | -------------:| :-----------:| :----------:| 
| **Buy HDTV     ** | yes | no | total |
| yes  | 105| 87 | 192 | 
| no | 40| 62 | 102 |   
| total | 145 | 149 | 294 | 
 

##### Table 3: Example of three-way contingency table
| | |   Buy Exercise machine |     |     |
|--- | :------------- | -------------:| :-----------:| :----------:| 
|**Customer group** | **Buy HDTV     ** | yes | no | total |
|College students | yes  | 2| 9 | 11 | 
| | no | 5| 20 | 25 |
| Working adults | yes  | 103| 78 | 181 | 
| | no | 35| 42 | 77 |  



Table 1:  TP = 105
          TN = 62
          FP = 87
          FN = 40
          Odds ratio = (TP/FP)/(FN/TN) = 1.8707
Table 2:  TP1 = 2
          TN1 = 20
          FP1 = 9
          FN1 = 5
          Odds ratio 1 = 0.8889
          TP2 = 103
          TN2 = 42
          FP2 = 78
          FN2 = 35
          Odds ratio 2 = 1.5846
          
If we combine the 2 groups into one table, it seems like the association is much stronger. The ratio is high, which means that TP, TN or both are high (relative to FP, FN and both respectively). So the association rule gets way stronger if the groups are merged. The direction of the association rule changes to a positive association, if one buys a HDTV there is a big chance of that person also buying an exercise machine.

Table 1:  phi's coefficient = TP * TN - FP * FN / sqrt(.P * .N * T. * F.)
                            = 105* 62 - 87 * 40 / sqrt(192*102 * 167*127)
                            = 0.1487
Table 2:  phi's coefficient = -0.01718
          phi's coefficient = 0.1056
          
The higher the absolute value of the phi's coefficient, the stronger the association. Again we see that when we combine the 2 groups into 1 larger group, the association gets stronger. The direction changes to a positive association, if one buys a HDTV there is a big chance of that person also buying an exercise machine.

Table 1:  interest factor  = P(Y|X)/P(Y) = TP / .P
                           = 0.5469
Table 2:  interest factor  = 0.1818
          interest factor  = 0.5659
          
The higher the interest factor, the stronger the association. This is the first measure where we don't see the positive effect on the association strength of combined groups. The  combined association is almost as strong as the association of the second group in table 2. This probably is because the first group in table 2 has few data points, so doesnt have as big of an impact on the total as the second group.